keras_cv/losses/dice.py

```diff
@@ -0,0 +1,215 @@
+                Defaults to 'dice'.
+        """
+        super().__init__(name=name, **kwargs)
+        self.beta = beta
+        self.from_logits = from_logits
+        self.loss_type = loss_type
+        self.label_smoothing = label_smoothing
+        self.per_sample = per_sample
+        self.epsilon = epsilon
+
+        self.axis = tf.constant(axis)
+        if not self.per_sample:
+            self.axis = tf.concat([tf.constant([0]), self.axis], axis=0)
+
+        if class_ids is not None:
+            if isinstance(class_ids, float) or any(
+                isinstance(x, float) for x in class_ids
+            ):
+                raise ValueError(
+                    f"The indices should be int or a list of integer. Got {class_ids}"
+                )
+            elif isinstance(class_ids, int):
+                class_ids = [class_ids]
+
+        if self.loss_type is not None:
+            if self.loss_type not in ["generalized", "adaptive"]:
+                raise ValueError(
+                    "The loss type is not valid. "
+                    f"It should be `generalized` or `adaptive`. Got {self.loss_type}"
+                )
+
+        self.class_ids = class_ids
+
+    def _smooth_labels(self, y_true, y_pred, label_smoothing):
+        num_classes = tf.cast(tf.shape(y_true)[-1], y_pred.dtype)
+        return y_true * (1.0 - label_smoothing) + (label_smoothing / num_classes)
+
+    def call(self, y_true, y_pred):
+        y_pred = tf.convert_to_tensor(y_pred)
+        y_true = tf.cast(y_true, y_pred.dtype)
+        label_smoothing = tf.convert_to_tensor(self.label_smoothing, dtype=y_pred.dtype)
+
+        if self.from_logits:
+            y_pred = tf.nn.softmax(y_pred)
+
+        if tf.cast(label_smoothing, dtype=tf.bool):
+            y_true = self._smooth_labels(y_true, y_pred, label_smoothing)
+
+        if self.class_ids is not None:
+            y_true, y_pred = gather_channels(y_true, y_pred, indices=self.class_ids)
+
+        # loss calculation: Fβ-score (in terms of Type I and type II erro
+        true_positive = keras.backend.sum(y_true * y_pred, axis=self.axis)
+        false_positive = keras.backend.sum(y_pred, axis=self.axis) - true_positive
+        false_negative = keras.backend.sum(y_true, axis=self.axis) - true_positive
+
+        power_beta = 1 + self.beta**2
+        numerator = power_beta * true_positive + self.epsilon
+        denominator = (
+            (power_beta * true_positive)
+            + (self.beta**2 * false_negative)
+            + false_positive
+            + self.epsilon
+        )
+
+        if self.loss_type == "generalized":
+            dice_score = self._generalized_dice_score(y_true, numerator, denominator)
+        elif self.loss_type == "adaptive":
+            dice_score = self._adaptive_dice_score(numerator, denominator)
+        else:
+            dice_score = numerator / denominator
+
+        if self.per_sample:
+            dice_score = keras.backend.mean(dice_score, axis=0)
```

Comment:
Keras loss should return scalar tensor, so per_sample result is bit anti-pattern here.